### PR TITLE
Merge v1.12.1

### PR DESCRIPTION
**Pull request by Tristan-Wilson**
Merge upstream geth's v1.12.1 release into our geth fork: https://github.com/ethereum/go-ethereum/releases/tag/v1.12.1

Merges were done with my gradual merging script to create a merge commit for the first conflict that is found: https://github.com/Tristan-Wilson/MyScripts/blob/main/bin/merge-helper. This makes for more merge commits but hopefully each one is easier to review.

Companion nitro PR: https://github.com/OffchainLabs/nitro/pull/1903

# Testing done
geth and nitro tests pass.

### CODE REVIEW

**Review by tsahee — COMMENTED** [review]
initial

**Review by tsahee — COMMENTED** [review]
tiny comments. great job on the merge commits

File: arbitrum/export.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 type TransactionArgs = ethapi.TransactionArgs
 
 func EstimateGas(ctx context.Context, b ethapi.Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap uint64) (hexutil.Uint64, error) {
-	return ethapi.DoEstimateGas(ctx, b, args, blockNrOrHash, gasCap)
+	return ethapi.DoEstimateGas(ctx, b, args, blockNrOrHash, nil, gasCap)
```
[the remark this answers is not in the document]
Comment (Tristan-Wilson) [reply arbitrum/export.go]: Fixed it

File: core/types/receipt.go [hunk 2]
```diff
@@ -368,18 +372,24 @@ func (rs Receipts) Len() int { return len(rs) }
...
+	}
+	w.WriteByte(r.Type)
 	switch r.Type {
-	case LegacyTxType, ArbitrumLegacyTxType:
+	case AccessListTxType, DynamicFeeTxType, BlobTxType:
 		rlp.Encode(w, data)
 	default:
-		w.WriteByte(r.Type)
+		// Arbitrum: same as above, but this should help prevent future merge conflicts,
+		// by keeping above case statement the same.
```
[the remark this answers is not in the document]
Comment (Tristan-Wilson) [reply core/types/receipt.go:385]: Done

File: core/state/statedb_arbitrum.go [hunk 3]
```diff
@@ -33,16 +33,16 @@ func (s *StateDB) GetUnexpectedBalanceDelta() *big.Int {
 	return new(big.Int).Set(s.unexpectedBalanceDelta)
 }
 
-func (s *StateDB) GetSuicides() []common.Address {
-	suicides := []common.Address{}
+func (s *StateDB) GetSelfDestructs() []common.Address {
```
[the remark this answers is not in the document]
Comment (Tristan-Wilson) [reply core/state/statedb_arbitrum.go:36]: I don't understand what you mean. Upstream removed the term Suicides so there is no function like that.

[the remark this answers is not in the document]
Comment (tsahee) [reply core/state/statedb_arbitrum.go:36]: you're right. must've gotten mixed up looking at stuff during review..

**Review by tsahee — APPROVED** [review]
LGTM

**Pull request merged**